Live grain migration

ReubenBond opened the pull request:
This PR implements live grain migration and resolves #7692

It contains only the mechanism for migration for now, no policies for automatic grain migration. The intention is to introduce those in separate PRs. The mechanism is accessible in two ways:

1. Within a grain, call `this.MigrateOnIdle()`
1. With a grain reference, call `grain.Cast<IGrainManagementExtension>().MigrateOnIdle()`

Migration supports preserving in-memory state without needing to refresh from storage. To access this functionality, grains *can* implement `IGrainMigrationParticipant` and components can inject `IGrainContext`, access `ObservableLifecycle`, and use `AddGrainMigrationParticipant`. I do not intend for grains to implement `IGrainMigrationParticipant` and the relevant interfaces are optimized for performance over usability. Instead, components of the grain which know how to preserve their state should implement the interface. The relevant interfaces here are:
``` csharp
public interface IGrainMigrationParticipant
{
    /// <summary>
    /// Called on the original activation when migration is initiated, before <see cref="IGrainBase.OnDeactivateAsync(DeactivationReason, CancellationToken)"/>.
    /// The participant can access and update the dehydration context.
    /// </summary>
    /// <param name="dehydrationContext"></param>
    void OnDehydrate(IDehydrationContext dehydrationContext);

    /// <summary>
    /// Called on the new activation after a migration, before <see cref="IGrainBase.OnActivateAsync(CancellationToken)"/>.
    /// The participant can restore state from the migration context.
    /// </summary>
    /// <param name="rehydrationContext">The rehydration context.</param>
    void OnRehydrate(IRehydrationContext rehydrationContext);
}

/// <summary>
/// Records the state of a grain activation for migration.
/// </summary>
public interface IDehydrationContext
{
    /// <summary>
    /// Gets the keys in the context.
    /// </summary>
    IEnumerable<string> Keys { get; }

    /// <summary>
    /// Adds a value to the dehydration context, associated with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value.</param>
    void Add(string key, ReadOnlySpan<byte> value);

    /// <summary>
    /// Adds a value to the dehydration context, associated with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="valueWriter">A delegate used to write the provided value to the context.</param>
    /// <param name="value">The value to provide to <paramref name="valueWriter"/>.</param>
    void Add(string key, Action<object, IBufferWriter<byte>> valueWriter, object value);
}

/// <summary>
/// Contains the state of a grain activation for migration.
/// </summary>
public interface IRehydrationContext
{
    /// <summary>
    /// Gets the keys in the context.
    /// </summary>
    IEnumerable<string> Keys { get; }

    /// <summary>
    /// Tries to get a value from the rehydration context, associated with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value, if present.</param>
    /// <returns><see langword="true"/> if the key exists in the context, otherwise <see langword="false"/>.</returns>
    bool TryGetValue(string key, out ReadOnlySequence<byte> value);
}
```

The lifecycle of a grain therefore becomes:

``` mermaid
stateDiagram-v2
    direction LR
    [*] --> Constructor
    Constructor --> OnActivateAsync
    Constructor --> OnRehydrate
    OnActivateAsync --> Running
    OnRehydrate --> OnActivateAsync 
    OnDeactivateAsync --> OnDehydrate
    Running --> OnDeactivateAsync 
    OnDeactivateAsync --> DisposeAsync
    OnDehydrate --> DisposeAsync
    DisposeAsync --> [*]
```

Note that rehydration occurs *before* `OnActivateAsync` is called and dehydration occurs *after* `OnDeactivateAsync` is called.

TODO before merge:
* [ ] Split PooledBuffer into a separate PR
* [ ] Add additional test coverage
* [ ] Implement `IGrainMigrationParticipant` for storage providers to preserve state

mitchdenny commented on test/Benchmarks/Ping/PingBenchmark.cs:
```diff
@@ -43,7 +44,9 @@ public PingBenchmark(int numSilos, bool startClient, bool grainsOnSecondariesOnl
                     }
-                });
+                })
+                    //.ConfigureLogging(logging => logging.AddConsole().SetMinimumLevel(LogLevel.Debug).AddFilter("Orleans.Runtime.Placement.PlacementService", LogLevel.Information))
```
Intentional?

ReubenBond replied:
No, thanks

mitchdenny commented on test/Benchmarks/Ping/PingBenchmark.cs:
```diff
@@ -68,6 +71,7 @@ public PingBenchmark(int numSilos, bool startClient, bool grainsOnSecondariesOnl
                     }
                 });
 
+                //hostBuilder.ConfigureLogging(logging => logging.AddConsole().SetMinimumLevel(LogLevel.Debug));
```
Intentional?

ReubenBond replied:
Fixed

benjaminpetit commented on src/Orleans.Core.Abstractions/GrainDirectory/IGrainDirectory.cs:
```diff
@@ -17,7 +17,7 @@ public interface IGrainDirectory
         /// <param name="address">The <see cref="GrainAddress"/> to register</param>
         /// <returns>The <see cref="GrainAddress"/> that is effectively registered in the directory.</returns>
-        Task<GrainAddress> Register(GrainAddress address);
+        Task<GrainAddress> Register(GrainAddress address, GrainAddress? previousAddress);
```
I think we can afford having breaking changes (I don't expect a lot of people having a custom directory...)

ReubenBond replied:
I was thinking of using a default interface method which falls back to unregister+register

benjaminpetit replied:
Having a directory that does not implement the "replace" atomically will produce some race conditions during migrations though. Not sure if it is really a big deal, but on some very active grains this migration could fail often.

ReubenBond replied:
The race is acceptable: if the destination grain loses the race, it drops the state and deactivate to migrate. It's the same as a regular activation race today. I am also happy for us to make a breaking change instead.

benjaminpetit commented on src/Orleans.Runtime/Catalog/Catalog.cs:
```diff
@@ -276,33 +281,51 @@ public void UnregisterSystemTarget(ISystemTarget target)
+            // Rehydration occurs before activation.
+            if (rehydrationContext is not null)
+            {
+                result.Rehydrate(rehydrationContext);
```
Should we surround this with a try/catch block? If the `Rehydrate` fails, I assume we don't want to keep this activation

benjaminpetit commented on src/Orleans.Runtime/Catalog/Catalog.cs:
```diff
@@ -248,14 +249,17 @@ public void UnregisterSystemTarget(ISystemTarget target)
         {
             if (TryGetGrainContext(grainId, out var result))
             {
+                rehydrationContext?.Dispose();
```
I think we should at least log a warning (maybe throw?) if `rehydrationContext` isn't null in this case

ReubenBond commented on src/Orleans.Core.Abstractions/Lifecycle/IGrainLifecycle.cs:
```diff
@@ -10,5 +16,78 @@ namespace Orleans.Runtime
+    /// <summary>
+    /// Records the state of a grain activation for migration.
+    /// </summary>
+    public interface IDehydrationContext
```
That sounds useful. In that case, maybe we should put IServiceProvider or IGrainContext on the IDe/RehydrationContext. I was thinking that this should mostly be used by grain storage and that should use the storage serializer.

ReubenBond replied:
Added in latest commit. Now, the interfaces look like this:
```C#
/// <summary>
/// Records the state of a grain activation which is in the process of being dehydrated for migration to another location.
/// </summary>
public interface IDehydrationContext
{
    /// <summary>
    /// Gets the keys in the context.
    /// </summary>
    IEnumerable<string> Keys { get; }

    /// <summary>
    /// Adds a sequence of bytes to the dehydration context, associating the sequence with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value.</param>
    void AddBytes(string key, ReadOnlySpan<byte> value);

    /// <summary>
    /// Adds a sequence of bytes to the dehydration context, associating the sequence with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="valueWriter">A delegate used to write the provided value to the context.</param>
    /// <param name="value">The value to provide to <paramref name="valueWriter"/>.</param>
    void AddBytes<T>(string key, Action<T, IBufferWriter<byte>> valueWriter, T value);

    /// <summary>
    /// Attempts to a value to the dehydration context, associated with the provided key, serializing it using <see cref="Orleans.Serialization.Serializer"/>.
    /// If a serializer is found for the value, and the key has not already been added, then the value is added and the method returns <see langword="true"/>.
    /// If no serializer exists or the key has already been added, then the value is not added and the method returns <see langword="false"/>.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value to add.</param>
    bool TryAddValue<T>(string key, T? value);
}

/// <summary>
/// Contains the state of a grain activation which is in the process of being rehydrated after moving from another location.
/// </summary>
public interface IRehydrationContext
{
    /// <summary>
    /// Gets the keys in the context.
    /// </summary>
    IEnumerable<string> Keys { get; }

    /// <summary>
    /// Tries to get a sequence of bytes from the rehydration context, associated with the provided key.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value, if present.</param>
    /// <returns><see langword="true"/> if the key exists in the context, otherwise <see langword="false"/>.</returns>
    bool TryGetBytes(string key, out ReadOnlySequence<byte> value);

    /// <summary>
    /// Tries to get a value from the rehydration context, associated with the provided key, deserializing it using <see cref="Orleans.Serialization.Serializer"/>.
    /// If a serializer is found for the value, and the key is present, then the value is deserialized and the method returns <see langword="true"/>.
    /// If no serializer exists or the key has already been added, then the value is not added and the method returns <see langword="false"/>.
    /// </summary>
    /// <param name="key">The key.</param>
    /// <param name="value">The value, if present.</param>
    /// <returns><see langword="true"/> if the key exists in the context, otherwise <see langword="false"/>.</returns>
    bool TryGetValue<T>(string key, out T? value);
}

ReubenBond commented on src/Orleans.Runtime/Catalog/ActivationData.cs:
```diff
@@ -391,6 +411,68 @@ private void ScheduleOperation(object operation)
+            var newLocation = await placementService.PlaceGrainAsync(GrainId, requestContext, PlacementStrategy);
+            if (newLocation == Address.SiloAddress || newLocation is null)
+            {
+                // No more appropriate silo was selected for this grain, but we cannot cancel migration now, so we must continue to deactivate and reactivate the grain.
```
This is the first thing which happens when migrating, so the cost is just the cost of running placement, which is cheap.
I've updated the comment, since this one is now inaccurate: the migration will be canceled, and the grain will not be deactivated.

ReubenBond replied:
I didn't answer your actual point: do we want to somehow prevent the grain code from spamming `MigrateOnIdle()`? Maybe. I'm not sure what the behavior should be. It seems like throwing wouldn't be useful. We could make the method async and return `ValueTask<bool>` or something else (with an implementation/design change). Something makes me uncomfortable about that. Another option is to ignore migration requests if they fall within some time period, but I think the added complexity is not worth it for the small benefit that you gain (not re-running placement, which is typically cheap).

oising replied:
> I didn't answer your actual point: do we want to somehow prevent the grain code from spamming `MigrateOnIdle()`? 

Would it make sense to throw if there's only a single silo?

ReubenBond commented on src/Orleans.Core.Abstractions/Lifecycle/IGrainLifecycle.cs:
```diff
@@ -10,5 +16,78 @@ namespace Orleans.Runtime
+        void RemoveMigrationParticipant(IGrainMigrationParticipant participant);
+    }
+
+    public interface IGrainMigrationParticipant
```
I'm ok with using different tenses. Can you think of a use case for using past/future tense instead of present tense?

To me, past/future tense make the interface seem more like it's for (passive) observers than for (active) participants. I could be in the minority. I also find it slightly confusing: as an implementer, what should I be doing in each of those methods? Should either of them be asynchronous?

Maybe the interface should be renamed to `IGrainLifecycleParticipant` or `IGrainLifecycleFilter` and it can be used as a place to add post-activation hooks, etc, satisfying #4519

As a counter argument (to my argument), you could take precedence from .NET BinaryFormatter which had OnSerializing/OnSerialized/OnDeserializing/OnDeserialized hooks - then again, those aren't the actual serialize/deserialize methods, they are there as some kind of pre/post processing. eg, to grab something from the serialization context after the fields have been deserialized in order to populate some value.

In Orleans, the grain base class has overridable `OnActivateAsync` and `OnDeactivateAsync` methods for the existing lifecycle events, which sets some precedence.

mitchdenny replied:
My issue with present tense is that it is almost never true. You either fire these callbacks before or after something happens and if it fires in the middle it usually makes a case for it being something more specific. I don't think we need both callback methods but I think its good to communicate WHEN something occurs before or after so that there is no doubt.

Fair point about the existing OnActivateAsync and OnDeactivateAsync. I think provided its documented WHEN it occurs it should be fine.

ReubenBond commented:
@mitchdenny & @benjaminpetit, please take another look

ReubenBond commented:
I believe I have addressed feedback adequately and have completed the original TODO items (largely, adding test coverage in various places). Please give this a final review and squash+merge when ready, @benjaminpetit

galvesribeiro commented:
Awesome! I see it being used on multiple use-cases, specially in game servers! :shipit:

benjaminpetit commented on src/Orleans.Core.Abstractions/Core/Internal/IGrainManagementExtension.cs:
```diff
@@ -12,6 +12,12 @@ public interface IGrainManagementExtension : IGrainExtension
+        /// Attempts to migrate the current instance to a new location once it becomes idle.
+        /// </summary>
+        /// <returns>A <see cref="Task"/> which represents the method call.</returns>
+        ValueTask MigrateOnIdle();
```
I wonder if we should call it `TryMigrateOnIdle` since it's not guaranteed, if people don't read the doc :p

ReubenBond replied:
Good question. It not returning a `bool` is my only issue, but I suppose we could make it async and make it return a `Task<bool>` or `Task<SiloAddress>` (prefer not this). Then again, would those be actionable results?

benjaminpetit replied:
Na it's fine for now imo

benjaminpetit commented on src/Orleans.Streaming/MemoryStreams/MemoryStreamQueueGrain.cs:
```diff
@@ -44,12 +45,25 @@ public Task<List<MemoryMessageData>> Dequeue(int maxCount)
+
+        void IGrainMigrationParticipant.OnDehydrate(IDehydrationContext dehydrationContext)
+        {
+            dehydrationContext.TryAddValue("queue", _eventQueue);
```
Not important right now: we should log when adding the _eventQueue failed, just to help debugging/catch some bugs
